### PR TITLE
[Commercial Server] New server config & server code

### PR DESCRIPTION
**Pull request by corey-nm**
Summary of changes:
- Add new EndpointConfig and ServerConfig objects
    - Removed integration field - this should be handled outside of these scripts
    - Added bucketing configuration
- Add conversion of EndpointConfig to PipelineConfig
- Move server app building logic into `server/server.py`
- Move cli logic into `server/cli.py`
- Split cli into two subcommands:
    - `task` which takes CLI options and a task and runs a server with a single model
    - `config` which only takes a config file
- endpoints for files are automatically added if `from_files` exists on input_schema
- `/invocations` is automatically added if `from_files` exists and there's only 1 model
- `Context` is now initialized with num_streams and scheduler set.
- Removed passing configuration through environment variables
- uvicorn hard coded to only use 1 worker
- setting ThreadPoolExecutor to have same number of workers as `Context.num_streams`
- Make model_path optional for `task` subcommand

Example configuration:
```yaml
num_cores: 16
num_workers: 8
endpoints:
  - name: main qa static model
    endpoint: /predict/static
    task: question-answering
    model: zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/12layer_pruned80_quant-none-vnni
    batch_size: 1
    bucketing:
      sequence_lengths: [10, 128]
```

Example commands:
```
deepsparse.server config --host 0.0.0.0 --port 8080 server-config.yaml
```
```
deepsparse.server task qa
```

### CODE REVIEW

File: src/deepsparse/server_v2/main.py
Comment (corey-nm) [inline src/deepsparse/server_v2/main.py:185]: This can be handled in another PR for commercial server. In theory we should be able to split up input that is divisible by batch size

File: src/deepsparse/server_v2/config.py
Comment (corey-nm) [inline src/deepsparse/server_v2/config.py:67]: This field can be removed since #604 adds dynamic batching for any batch size

File: tests/test_server/test_endpoints.py
Comment (rahul-tuli) [inline tests/test_server/test_endpoints.py:160]: Remove or add back in?

File: tests/test_server/test_endpoints.py
Comment (rahul-tuli) [inline tests/test_server/test_endpoints.py:200]: remove or add back in?

File: src/deepsparse/engine.py
Comment (corey-nm) [reply src/deepsparse/engine.py:602]: Good idea!

File: tests/test_server/test_endpoints.py
Comment (corey-nm) [reply tests/test_server/test_endpoints.py:160]: Oh yeah thanks for reminding me, I wanted to add more tests here. Ideally we could get the mock config PR before this so they aren't super slow.

File: tests/test_server/test_endpoints.py
Comment (corey-nm) [reply tests/test_server/test_endpoints.py:200]: Will add back in thanks

File: src/deepsparse/server/cli.py
Comment (corey-nm) [reply src/deepsparse/server/cli.py:129]: I just tried, it actually does already support this thanks to that context thing we added before!!

File: tests/test_server/test_endpoints.py
Comment (corey-nm) [reply tests/test_server/test_endpoints.py]: Okay so I'm taking out bucketing & dynamic batch testing at this server level. dynamic batch test is below, but will require the #604 PR to be merged. Bucketing seems to have some issues so we can address that and add tests in a separate PR.

```python
                EndpointConfig(
                    name="test endpoint 3",
                    endpoint="/predict/bucketed",
                    task="text-classification",
                    model=stub,
                    batch_size=1,
                    bucketing=SequenceLengthsConfig(sequence_lengths=[2, 4]),
                ),

    def test_dynamic_batch_handles_any_batch_size(self, client):
        sentences = [
            "today is great",
            "today is terrible",
            "today is average",
            "just okay today, thanks",
            "is this thing working??",
        ]
        for num_items in range(1, len(sentences) + 1):
            response = client.post(
                "/predict/dynamic-batch",
                json={"sequences": sentences[:num_items]},
            )
            assert response.status_code == 200
            output = response.json()
            assert len(output["labels"]) == num_items
            assert len(output["scores"]) == num_items
```

Comment (corey-nm) [reply tests/test_server/test_endpoints.py]: Added a task in asana for this

File: src/deepsparse/engine.py
Comment (bfineran) [inline src/deepsparse/engine.py:575]: mentioned offline - we should only be supporting elastic scheduler with `Context`

File: src/deepsparse/pipeline.py
Comment (bfineran) [inline src/deepsparse/pipeline.py:302]: what's the intended usage for this? note with auto install it may have some unintended side effects like installing the nm-transformers fork a bit early. if it's just to enable optional `model_path` in the cli, let's support a `"default"` special string that will let us get through the config schema

File: src/deepsparse/server/server.py
Comment (bfineran) [inline src/deepsparse/server/server.py:73]: should remove and require elastic

File: src/deepsparse/pipeline.py
Comment (bfineran) [inline src/deepsparse/pipeline.py:697]: this is fine in controlled environments for getting the best single model numbers - also since we're always overriding with a context (elastic) in the server. however might want to think about how we can detect misuse and warn (ie initiating multiple single stream pipelines that are intended to run asynchronously )

File: src/deepsparse/server/config.py
Comment (bfineran) [inline src/deepsparse/server/config.py:124]: let's add an `is_cv` helper, probably won't be the last time we use this if statement / add another cv pipeline

File: src/deepsparse/server/config.py
Comment (bfineran) [inline src/deepsparse/server/config.py:82]: we should set the pipeline `alias` to the `name` or `endpoint` value here, that way the logger can have something easy to log the pipeline name as

File: src/deepsparse/server/server.py
Comment (bfineran) [inline src/deepsparse/server/server.py:97]: the logic for re-routing the endpoint path with /invocations for sagemaker support was taken out. need to get some functionality for this added back in and update the existing sagemaker documentation/dockerfile if the location of where to set integration has changed. ping if any questions

File: src/deepsparse/server/server.py
Comment (bfineran) [inline src/deepsparse/server/server.py:153]: we need `invocations` for situations besides from_files, such as nlp which doesn't currently support from_files

File: src/deepsparse/server/server.py
Comment (bfineran) [inline src/deepsparse/server/server.py:137]: will using the same threadpool for the run loop as well as inside the pipeline (ie for dyamic) give us an inherent "minus 1" on concurrent inferences since one worker will be dedicated to this main predict func?

**Review by bfineran — CHANGES_REQUESTED** [review]

**Comment by bfineran** [comment]
additionally, docs seem fine now (outside of sagemaker) in terms of maintianing existing functionality. but should start thinking about how we want to structure docs around configuring for features such as bucketing, dynamic, custom w/ server, etc

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py:153]: Gotcha, I was translating the logic from the existing server, which only allows file uploads with sagemaker. You okay with doing sagemaker cleanup in a separate PR?

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py:137]: Hmm good question. Unclear how this all works with multiple concurrent server requests. We may also just be able to remove this async run_in_executor thing and do `pipeline(request)`. I think this needs some benchmarking to say.

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py:97]: Can you expand? The server will add an /invocations endpoint if there's only 1 endpoint, was there something else I missed bringing over?

File: src/deepsparse/pipeline.py
Comment (corey-nm) [reply src/deepsparse/pipeline.py:697]: I can revert these changes from the PR - don't set the scheduler field of PipelineConfig anymore

File: src/deepsparse/pipeline.py
Comment (corey-nm) [reply src/deepsparse/pipeline.py:302]: Yeah just for optional model path, good idea

File: src/deepsparse/server/server.py
Comment (bfineran) [reply src/deepsparse/server/server.py:153]: surprising, our advertised examples use the nlp pipelines which don't have file support IIRC, maybe this was a late introduction post the last time we tested sagemaker

File: src/deepsparse/server/server.py
Comment (bfineran) [reply src/deepsparse/server/server.py:97]: missing functionality expanded on in thread below. and would be good to add/keep in a pathway for specifying deployment scenario (can be above the server config level, like CLI) because for something like multi-model sagemaker we'll want to know and detect when to add in the extra sagemaker routes. having knowledge of integrations is also important for choosing which logger integration(s) to use/add in

additionally, don't think it's necessary to always add in `invocations` for single model servers since for non sagemaker deployments it will just be extra spam in the auto docs/swagger ui

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py:153]: Ahh I was interpretting code incorrectly - nevermind!

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py]: Okay I added integration field back in, and now the logic for sagemaker should match what was there before

File: src/deepsparse/server/config.py
Comment (corey-nm) [reply src/deepsparse/server/config.py:138]: Can we do this in another PR? don't want to add an additional file to this PR to review

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py]: I just ran some benchmarking, and this version performed the same as just calling `pipeline(request)`, so I replaced it to not use the async version.

File: src/deepsparse/engine.py
Comment (bfineran) [reply src/deepsparse/engine.py:575]: we should just remove this, no?

Comment (corey-nm) [reply src/deepsparse/engine.py:575]: Oh yeah - will revert now

File: src/deepsparse/server/config.py
Comment (bfineran) [reply src/deepsparse/server/config.py:138]: it should be a minimal change with basically this same code no?  Just worried about this falling through the cracks. open an asana ticket if you feel strongly

File: src/deepsparse/server/server.py
Comment (bfineran) [inline src/deepsparse/server/server.py:97]: was removed - let's make sure to find any other usages of this kwarg

**Review by bfineran — COMMENTED** [review]
lgtm overall pending comment

File: src/deepsparse/server/config.py
Comment (corey-nm) [reply src/deepsparse/server/config.py]: Added

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py]: Updated - no other usages

File: .github/workflows/test-check.yaml
Comment (dbogunowicz) [inline .github/workflows/test-check.yaml:34]: I think it would be cleaner to separate rudimentary deepsparse tests from server tests, especially now when the codebase for the server will grow quickly. Not a blocking comment, but something to think about.

File: src/deepsparse/server/server.py
Comment (dbogunowicz) [inline src/deepsparse/server/server.py:70]: Could you elaborate? What would happen if `workers=2`?

File: src/deepsparse/server/server.py
Comment (dbogunowicz) [inline src/deepsparse/server/server.py:125]:
```suggestion
    # create pipelines & endpoints
```

File: src/deepsparse/server/main.py
Comment (dbogunowicz) [inline src/deepsparse/server/main.py:7]: Are we sure that landing that PR will be identical to replacing the old server? Maybe it makes sense to have them both live concurrently (something like server and server_v2) for a while.

File: src/deepsparse/server/config.py
Comment (dbogunowicz) [inline src/deepsparse/server/config.py:75]: nit: not `Union[Optional[ImageSizesConfig], Optional[SequenceLenghtsConfig]]?

File: src/deepsparse/server/config.py
Comment (dbogunowicz) [inline src/deepsparse/server/config.py:135]: Good call on unifying the set of cv tasks.

File: src/deepsparse/pipeline.py
Comment (dbogunowicz) [inline src/deepsparse/pipeline.py:179]: I understand that this predates the `unify-run-split-join` branch and will be updated once the branch lands?

**Review by dbogunowicz — COMMENTED** [review]
Great job, very clean implementation.
Left a few nits, ping me aggressively for further reviews.

File: src/deepsparse/pipeline.py
Comment (corey-nm) [reply src/deepsparse/pipeline.py:179]: Yes, though I believe we'll be landing this PR first, will make sure its updated either way

File: src/deepsparse/server/server.py
Comment (corey-nm) [reply src/deepsparse/server/server.py:70]: num_workers for uvicorn or other server things usually means multiple processes handling requests. In our case this is bad because *I think* each server process would have it's own copy of context/engine. Whereas we want to ensure that there's only 1 context/engine

File: src/deepsparse/server/main.py
Comment (corey-nm) [reply src/deepsparse/server/main.py:7]: Yeah good question. I think this is another big reason to have unit/integration tests, otherwise all you can *really* do is try to side-by-side compare things haha. Are there specific cases that you are wondering about?

File: .github/workflows/test-check.yaml
Comment (corey-nm) [reply .github/workflows/test-check.yaml:34]: Yeah I could even see the server being in a separate repo

File: src/deepsparse/server/config.py
Comment (corey-nm) [reply src/deepsparse/server/config.py:75]: I prefer having the optional on the outside because the whole field is optional. Otherwise it could be valid to have a non-optional field in the union (e.g. `Union[Optional[ImageSizesConfig], SequenceLengthsConfig]`), which is not what the intention is.

**Comment by bfineran** [comment]
this command needs update as well (might be good to do a code search on invocations of the server in general): https://github.com/neuralmagic/deepsparse/blob/8b3426d2ff2551b0c07faad1a8c79797d314b633/examples/aws-sagemaker/Dockerfile#L27

**Comment by bfineran** [comment]
LGTM after docs update, merge conflict, and some simple benchmarks added to PR description. Great work @corey-nm

File: src/deepsparse/pipeline.py
Comment (rahul-tuli) [inline src/deepsparse/pipeline.py:372]: nice!

File: src/deepsparse/server/config.py
Comment (rahul-tuli) [inline src/deepsparse/server/config.py:164]: nit:
```suggestion
        # NOTE: convert from List[Tuple[int]] to List[List[int]]
```

File: src/deepsparse/tasks.py
Comment (rahul-tuli) [inline src/deepsparse/tasks.py:174]: Python can be so beautiful

**Review by rahul-tuli — APPROVED** [review]
Approved! Nice tests.

File: src/deepsparse/tasks.py
Comment (corey-nm) [reply src/deepsparse/tasks.py:174]: there's nothing better than writing some beautiful code 😁

**Review by bfineran — APPROVED** [review]
LGTM @corey-nm quick thing to note / get in - we standardize to putting `--help` text(s) at the top docstring of the python files where their scripts are created (ie for the `cli.py` file)

**Pull request merged**
